### DISCUSSION TITLE
New serialization framework can block delegating to existing implementations

### DISCUSSION THREAD

**Discussion by Ten0** [diesel-rs/diesel#2984] [Ideas]
As of #2931, we often can't delegate `ToSql` implementations to existing implementations anymore, e.g.:
```rust
struct SerializeAsi64 {
    val: u64,
}
impl ToSql<BigInt, Pg> for SerializeAsi64 {
    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Pg>) -> Result {
        let v: i64 = self.val.try_into()?;
        <i64 as ToSql<BigInt, Pg>>::to_sql(&v, out)
    }
}
```
```
error[E0597]: `v` does not live long enough
  --> src/lib.rs:29:44
   |
27 |     fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Pg>) -> Result {
   |               -- lifetime `'b` defined here
28 |         let v: i64 = self.val.try_into()?;
29 |         <i64 as ToSql<BigInt, Pg>>::to_sql(&v, out)
   |         -----------------------------------^^------
   |         |                                  |
   |         |                                  borrowed value does not live long enough
   |         argument requires that `v` is borrowed for `'b`
30 |     }
   |     - `v` dropped here while still borrowed
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6bca00f36aaa6635c87e4a8f8b38ab1e))

I'm wondering whether we'd like to do something to allow it, like this:
```rust
pub trait ToSql<'r, 'b, A, DB> {
    /// See the trait documentation.
    fn to_sql(&'r self, out: &mut Output<'b, '_, DB>) -> Result;
}

impl<'r, 'b, 'a: 'b, A, T, DB> ToSql<'r, 'b, A, DB> for &'a T
where
    T: ToSql<'a, 'b, A, DB> + ?Sized,
{
    fn to_sql(&'r self, out: &mut Output<'b, '_, DB>) -> Result {
        (*self).to_sql(out)
    }
}

impl<'r, 'b> ToSql<'r, 'b, BigInt, Pg> for i64 {
    fn to_sql(&'r self, out: &mut Output<'b, '_, Pg>) -> Result {
        out.write_val(*self)
    }
}
impl<'b, 'r: 'b> ToSql<'r, 'b, BigInt, Pg> for str {
    fn to_sql(&'r self, out: &mut Output<'b, '_, Pg>) -> Result {
        out.write(self)
    }
}
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=07d2619e371d520fc9916cb7a9d44e6b))

or like this:
```rust
pub trait ToSqlStatic<A, DB> {
    fn to_sql(&self, out: &mut Output<'_, '_, DB>) -> Result;
}
impl<T, A, DB> ToSql<A, DB> for T
where
    T: ToSqlStatic<A, DB>,
{
    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, DB>) -> Result {
        <T as ToSqlStatic<A, DB>>::to_sql(self, out)
    }
}
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b99204fe2871dd21e1483634e8c3da21))

## Workaround
<s>My current workaround is to copy-paste from the Diesel implementation:</s>
EDIT: If backend's `BindCollector` is `RawBytesBindCollector`, using [`reborrow`](https://docs.diesel.rs/master/diesel/serialize/struct.Output.html#method.reborrow) works:
```rust
impl<DB> diesel::serialize::ToSql<diesel::sql_types::Int8, DB> for CustomType
where
	DB: diesel::backend::Backend<BindCollector = diesel::query_builder::bind_collector::RawBytesBindCollector<DB>>,
	i64: diesel::serialize::ToSql<diesel::sql_types::Int8, DB>,
{
	fn to_sql<'b>(&'b self, out: &mut diesel::serialize::Output<'b, '_, DB>) -> diesel::serialize::Result {
		let n: i64 = self.inner.get().try_into()?;
		diesel::serialize::ToSql::<diesel::sql_types::Int8, DB>::to_sql(&n, &mut out.reborrow())
	}
}
```

**Comment by weiznich** [comment]
You can use [`Output::reborrow`](https://docs.diesel.rs/master/diesel/serialize/struct.Output.html#method.reborrow) to shorten the output lifetime for the inner `to_sql` call.
This is something that should be mentioned in the migration guide + maybe we should add an example for this to the insertable docs.

**Reply by toxeus** [reply]
@weiznich `reborrow()` isn't implemented for Sqlite (`BindCollector = SqliteBindCollector`). Is this an oversight or is there a limitation blocking this?

**Reply by weiznich** [reply]
@toxeus It's expected that `reborrow` is not available for the sqlite backend as the changes where implemented for that backend specifically. Sqlite allows to use borrowed pointers instead of copying all values. This improves the performance in some scenarios by a significant factor (see the linked PR for exact numbers). On the other hand that requires that the bound value lives as least as long till the query result in available, which is effectively exactly that what is expressed by the lifetime bounds there. This implies you cannot easily use a temporary value there, that is produced in the scope of `ToSql::to_sql`.

That all written: Our [internal implementation](https://github.com/diesel-rs/diesel/blob/8a53cb7c8a09ae891df6c104d1c8a03d51ee07fc/diesel/src/serialize.rs#L99-L109) supports using "temporary" values there by storing them as long as needed. We just don't expose this yet, as at least I consider that a internal implementation detail.  If there is really a need for having this functionality there we can talk about adding a corresponding interface for this. I'm open for suggestions here. To be clear here: Just making those functions `pub` is probably not the way to go. I would like to hear some possible alternatives here first.

**Reply by toxeus** [reply]
@weiznich thank you for your detailed answer. I do appreciate that performance is also a priority in diesel.

> This implies you cannot easily use a temporary value there, that is produced in the scope of `ToSql::to_sql`.

This is quite a breaking change for me. I have a lot of `ToSql:to_sql` implementations that rely on a temporary value for creating the database serialization similar how it's implemented for the datetime types.

>  If there is really a need for having this functionality there we can talk about adding a corresponding interface for this.

I think that there is definitely a need for supporting temporary values. The implementations for the datetime types are a proof for that need.

>  I'm open for suggestions here. To be clear here: Just making those functions pub is probably not the way to go. I would like to hear some possible alternatives here first.

I agree, making those functions pub is not ideal. Before, it was possible to generically implement `ToSql::to_sql()` with temporary values for any database backend. Simply exposing those functions will not restore that behavior. It would be great to be able to generically implement `ToSql::to_sql()` for any backend. That's why I asked about `reborrow()` for Sqlite. Using a smart pointer like `Cow` also yields to the same issues. Looks like a dead end to me.

**Reply by weiznich** [reply]
> This is quite a breaking change for me. I have a lot of ToSql:to_sql implementations that rely on a temporary value for creating the database serialization similar how it's implemented for the datetime types.

While this is unfortunate, it's at least expected. Thats the reason why there will be a 2.0 release and why I'm telling people not to use the "master" branch as "stable" target.

> I agree, making those functions pub is not ideal.

Happy to see proposals here. Otherwise this is something I do not consider blocking for the 2.0 release as I believe whatever interface we come up here can be later added on the top of the existing interface. A good starting point for such a discussion would be to collect cases where a temporary value seems to be necessary. From the refactoring of diesel itself it seems like most cases can be actually rewritten to support all backends with the new framework. That requires knowledge of the exact cases.

>  It would be great to be able to generically implement ToSql::to_sql() for any backend. That's why I asked about reborrow() for Sqlite. Using a smart pointer like Cow also yields to the same issues. Looks like a dead end to me.

To be clear here: That will likely never work again as the backends are just really different in how they handle binding values. The old sqlite implementation hide that from you by doing really nasty things that involved coping around values at least 3 times. That was really error prone and could lead to performance issues in some situations. The upside was that it looked like it "worked" the same way for outside users. Changing that was always something on our todo list. The 2.0 release seemed like a good opportunity to just do that.

**Reply by Ten0** [reply]
`reborrow` works fine for my purposes because I'm only on PG and unlikely to use SQLite any time soon. Thanks! :)
However it doesn't work for the same kind of implementations with SQLite.

>  If there is really a need for having this functionality there we can talk about adding a corresponding interface for this. I'm open for suggestions here. To be clear here: Just making those functions pub is probably not the way to go. I would like to hear some possible alternatives here first.

I have two in [my post above](https://github.com/diesel-rs/diesel/discussions/2984#discussion-3749047). :)
The first one would probably look cleaner once we have GAT.

**Reply by weiznich** [reply]
@Ten0  To be the critic person here:

> ```rust
>pub trait ToSql<'r, 'b, A, DB> {
>    /// See the trait documentation.
>    fn to_sql(&'r self, out: &mut Output<'b, '_, DB>) -> Result;
>}
>
>impl<'r, 'b, 'a: 'b, A, T, DB> ToSql<'r, 'b, A, DB> for &'a T
>where
>    T: ToSql<'a, 'b, A, DB> + ?Sized,
>{
>    fn to_sql(&'r self, out: &mut Output<'b, '_, DB>) -> Result {
>        (*self).to_sql(out)
>    }
>}
>
>impl<'r, 'b> ToSql<'r, 'b, BigInt, Pg> for i64 {
>    fn to_sql(&'r self, out: &mut Output<'b, '_, Pg>) -> Result {
>        out.write_val(*self)
>    }
>}
>impl<'b, 'r: 'b> ToSql<'r, 'b, BigInt, Pg> for str {
>    fn to_sql(&'r self, out: &mut Output<'b, '_, Pg>) -> Result {
>        out.write(self)
>    }
>}
>``` 

This would likely work to fix the problem with `ToSql`, but I see other problems here:
* It's at least unclear for me if this would be usable in the whole query building framework without requiring `'r: 'b` at some point. Likely not. In that case we would continue to have basically the same problem that it's only possible to use impls that actually fulfill this contract. 
* The "special trick" here seems to be the addition of `write_val` and `write`. That's not that different then just calling `.reborrow()` + the internal implementation from my point of view. For sqlite we need to have something different anyway as there is no concept like `write` there.
* Adding lifetimes to the trait itself is infectious. This would require adding lifetimes to basically all of diesels query builder traits + it would require having lifetimes in basically all generic code that down stream users will try to write. As this would make it even harder (or basically impossible) to write generic code involving diesel it would try to avoid this if possible.

>```rust
>pub trait ToSqlStatic<A, DB> {
>    fn to_sql(&self, out: &mut Output<'_, '_, DB>) -> Result;
>}
>impl<T, A, DB> ToSql<A, DB> for T
>where
>    T: ToSqlStatic<A, DB>,
>{
>    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, DB>) -> Result {
>        <T as ToSqlStatic<A, DB>>::to_sql(self, out)
>    }
>}
>```

That would require a lot more additional trait impls for all backends. Additionally implementing a custom type becomes more work. Although it would likely solve this specific problem.

<hr>

I've played in the last days a bit with updating [diesel-oci](https://github.com/gigainfosystems/diesel-oci) to incooperate the bind system rewrite. It turns out that this change is actually useful there and it seems like it will use a similar bind system than sqlite. That's a indicator for me that the new system is flexible enough to handle different usecases. That written I've noticed a few missing pieces to actually make this work:
* There is no way to deconstruct an `Output` later on from a third party crate. We may want to expose `Output::into_inner()` for this for third party backend crates.
* There is no way to set/access the value of the inner buffer if it does not implement `Write`. This might be required for custom backends that will bring their own `Value` enum.

Based on this situation I would like to summarize the current problem space here:

* The bind system rewrite did slightly change the requirements for implementing `ToSql`. It is not easily possible anymore to use temporary values as part of the `ToSql` impl. That's a potential problem for people upgrading their codebase as this is a "common" patterns for diesel 1.x `ToSql` impls. This implies we need to provide at least usable alternatives that are documented in the right places.
* For the `Mysql`/`Postgresql` backend (which use `RawBytesBindCollector`) there is the `Output::reborrow()` method which effectively fixes this issue. We need to make it more obvious that this methods solves problems with temporary values constructed inside of `ToSql::to_sql` by maybe adding an explicit example to `ToSql::to_sql` + adding a paragraph about this to the migration guide.
* For the `Sqlite` backend there is currently no way to construct and use a temporary value as part of the `ToSql::to_sql` implementation. Internally we solve this by using various private `set_x` functions that allow to set the corresponding value. We don't want to expose this functions. Additionally the `Sqlite` needs to differentiate between temporary and borrowed values. 
* Third party backends do not really have the possibility yet to provide a custom buffer type for the `BindCollector` impl, as there is no way to access the buffer from the outside. They may have similar requirements as the `Sqlite` backend on how bind values should be restricted.

To solve those issues I propose the following solution:

* We add the following function to `Output`:
  ```rust
    pub fn set_value<V>(&mut self, value: V)
    where
        V: Into<<crate::backend::BindCollector<'a, DB> as BindCollector<'a, DB>>::Buffer>,
    {
        self.out = value.into();
    }
  ```
* We wrap `SqliteBindValue<'a>` into a struct to make the enum opaque for outside users and provide a set of `From<_>` impls for the newly exposed wrapper. That's the only way to construct the "enum".
* We replace any usage of `Output::set_x()` internally with the newly created `set_value` function
* (We expose `Output::into_inner` for third party backend)

This allows:
* Postgres/Mysql users can use `reborrow()` to "fix" the lifetime for their temporary values
* Sqlite users to construct and use temporary values as part of their `ToSql::to_sql` implementation
* It provides us an extensible way to construct the internal bind value enum
* Third party backends can either use the write based interface or provide their own buffer type + corresponding `From<_>` impls
* It does not introduce additional traits or lifetimes

Comments (and PR's) welcome :wink:

**Reply by weiznich** [reply]
I've opened https://github.com/diesel-rs/diesel/pull/2998 to address this

**Comment by toxeus** [comment]
> While this is unfortunate, it's at least expected. Thats the reason why there will be a 2.0 release and why I'm telling people not to use the "master" branch as "stable" target.

Good to know that this is expected. It is my deliberate choice to use master for two reasons:

1. I'm depending on some of the 2.0 features
2. I want to track breaking changes such that I can discuss/report them before the release, e.g. [this one](https://github.com/diesel-rs/diesel/issues/2898)

I'm fully aware that by doing that I need to deal with changing APIs. But I'd need to do that too if I would use 1.X releases and then migrate to 2.0 once it's released but without the possibility to migrate gradually and without being able to understand the changes before they are released and therefore stable.

> Otherwise this is something I do not consider blocking for the 2.0 release as I believe whatever interface we come up here can be later added on the top of the existing interface.

That is my feeling too.

> A good starting point for such a discussion would be to collect cases where a temporary value seems to be necessary.

I'll try to find solutions for the existing `ToSql` impls and will report those use-cases that aren't possible to implement with the current API.

> To be clear here: That will likely never work again as the backends are just really different in how they handle binding values.

Thanks for the clarification. If that's a deliberate choice then this is perfectly fine for me.

**Comment by weiznich** [comment]
> I'm fully aware that by doing that I need to deal with changing APIs. But I'd need to do that too if I would use 1.X releases and then migrate to 2.0 once it's released but without the possibility to migrate gradually and without being able to understand the changes before they are released and therefore stable.

Thanks for this clarification :heart:. Such feedback on API changes is always welcome. 
One outcome of this discussion is that I noticed that we should have a paragraph about this in our planed migration guide, so that users know how to handle this situation. This should probably just point to `Output::reborrow` for the postgres and mysql backend and outline the underlying issue for the sqlite backend. Contributions here are welcome, the Changelog already contains fragments of such a guide.

**Reply by toxeus** [reply]
I'll try to contribute that section to the guide. Hopefully, I can get around doing that before the Christmas break. 

> our planed migration guide

I conclude that document doesn't exist yet. Should it be created in `./guide_drafts`?

**Reply by weiznich** [reply]
There are already a few fragments in our changelog: https://github.com/diesel-rs/diesel/blob/master/CHANGELOG.md?plain=1#L255-L332
Writing a "real" migration guide based on this is something that needs to be done before the release. I will likely start working on this as soon as we have released an 2.0.0-rc0, which hopefully happens next month or so.

**Comment by ensc** [comment]
What is the recommended way to write a generic implementation of `to_sql()` now?   Previously it was possible to write a DB agnostic one. Now, it seems that there must be written one for postgres stylish backends and one for sqlite like ones.

I have a lot of code like

```rs
#[diesel(sql_type = diesel::sql_types::SmallInt)]
pub enum FileType {
    SSTATE  = 1,
    SOURCE  = 2,
    SCM     = 3,
    CICACHE = 4,
}

impl <DB> ToSql<SmallInt, DB> for FileType
where
    DB: Backend,
    i16: ToSql<SmallInt, DB>
{
    fn to_sql<W: Write>(&self, out: &mut Output<W, DB>) -> SerResult {
	(*self as i16).to_sql(out)
    }
}
```

Changing this to new api does not seem to be trivial (the "Beside of the changed signature, existing impls of this trait should remain unchanged in almost all cases" in the [ChangeLog](https://github.com/diesel-rs/diesel/blob/bdd008c97b53957329ef2cae0504739e76dd3860/CHANGELOG.md?plain=1#L195-L196) definitively does not hold ;)

**Reply by weiznich** [reply]
> What is the recommended way to write a generic implementation of to_sql() now? Previously it was possible to write a DB agnostic one. Now, it seems that there must be written one for postgres stylish backends and one for sqlite like ones.

You have two options here:

* Either write a specific impls per bind collector (so one for sqlitr and one for postgres/mysql) and use the apis provided there to handle temporary values. That's [`Output::set_value`](https://docs.diesel.rs/master/diesel/serialize/struct.Output.html#method.set_value) for sqlite to set the corresponding value and [`Output::reborrow`](https://docs.diesel.rs/master/diesel/serialize/struct.Output.html#method.reborrow) for the other backends to shorten the lifetime so that you can call `to_sql` on a temporary value.
* Or you reorganize your `to_sql` implementation in such way that you dont produce a temporary value there. This means your cannot do the cast anymore to convert your enum to an integer. You could use an explicit mapping there for example.

> Changing this to new api does not seem to be trivial (the "Beside of the changed signature, existing impls of this trait should remain unchanged in almost all cases" in the ChangeLog definitively does not hold ;)

I would not say that this statement is completely wrong. That written I agree that we should probably weaken it so that it says thats the case for some database specific impls. For more complex cases the migration guide should likely describe the two strategies outlined above. Mind opening a PR for that?

**Reply by ensc** [reply]
> Either write a specific impls per bind collector ... That's `Output::set_value` for sqlite

I played with this, but problem is, that `set_value()` accepts only a very limited amount of types and does not use existing `ToSql` conversions.  E.g. one of the objects which I want to handle uses `chrono::naive::NaiveDateTime`:

```rs
pub struct Time(NaiveDateTime);
```

Deserialization is trivial:

```rs
impl <DB> FromSql<Timestamp, DB> for Time
where
    DB: Backend,
    NaiveDateTime: FromSql<Timestamp, DB>
{
    fn from_sql(bytes: diesel::backend::RawValue<DB>) -> DeserResult<Self> {
	let v = NaiveDateTime::from_sql(bytes)?;
	Ok(v.into())
    }
}
```

But serialization 

```rs
impl ToSql<Timestamp, diesel::sqlite::Sqlite> for Time
{
    fn to_sql(&self, out: &mut Output<diesel::sqlite::Sqlite>) -> SerResult {
        // formerly:  self.as_ref().to_sql(out)
	out.set_value(self.as_ref());
	Ok(diesel::serialize::IsNull::No)
    }
}
```

does not work because conversion into `SqliteBindValue` is only implemented for a handful of core types.



> Or you reorganize your to_sql implementation in such way that you dont produce a temporary value there

This is not possible, because there are non-enumerable types

**Reply by weiznich** [reply]
I do not see any reason why `Time::as_ref` cannot return a reference bound to the lifetime of self in your example. That would allow you to just call the corresponding `ToSql` impl for `NaiveDateTime`.

> I played with this, but problem is, that set_value() accepts only a very limited amount of types and does not use existing ToSql conversions. 

Thats by design and that are the types that are accepted by sqlite itself natively. We do not plan to extend the set of supported types there.

Generally speaking: I understand that this change can cause troubles. At least for me the performance and code simplification / safety improvements indicate that this is the right solution to the underlying problem. That means there wont be any fundamental changes to this feature before the next release. What is definitively possible is to improve rough corners like the migration guide, the documentation itself or even a few helper functions. Here we are happy to receive contributions.

**Reply by ensc** [reply]
> I do not see any reason why Time::as_ref cannot return a reference bound to the lifetime of self in your example. That would allow you to just call the corresponding ToSql impl for NaiveDateTime.

```rs
#[cfg(feature = "sqlite")]
impl ToSql<Timestamp, diesel::sqlite::Sqlite> for Time
{
    fn to_sql(&self, out: &mut Output<diesel::sqlite::Sqlite>) -> SerResult {
	ToSql::<Timestamp, diesel::sqlite::Sqlite>::to_sql(self.as_ref(), out)
    }
}
```

results in

```
32 |     fn to_sql(&self, out: &mut Output<diesel::sqlite::Sqlite>) -> SerResult {
   |               -----            ------------------------------
   |               |
   |               these two types are declared with different lifetimes...
33 |     ToSql::<Timestamp, diesel::sqlite::Sqlite>::to_sql(self.as_ref(), out)
   |                                                                       ^^^ ...but data from `self` flows into `out` here
```

**Reply by weiznich** [reply]
As long as you dont include all relevant code in your questions thats not answerable.

**Reply by ensc** [reply]
> As long as you dont include all relevant code in your questions thats not answerable.

what do you need more?  `as_ref()` returns `&self.0` (which can be used directly when `.0` is `pub`) and lifetime is this of `&self`.  So, to make it it complete: how do I migrate from

```rs
pub struct Time(pub NaiveDateTime);

impl <DB> FromSql<Timestamp, DB> for Time
where
    DB: Backend,
    NaiveDateTime: FromSql<Timestamp, DB>
{
    fn from_sql(bytes: diesel::backend::RawValue<DB>) -> DeserResult<Self> {
	let v = NaiveDateTime::from_sql(bytes)?;
	Ok(v.into())
    }
}

impl <DB> ToSql<Timestamp, DB> for Time
where
    DB: Backend,
    NaiveDateTime: ToSql<Timestamp, DB>
{
    fn to_sql<W: Write>(&self, out: &mut Output<W, DB>) -> SerResult
        self.0.to_sql(out)
    }
}
```

to the new API?

For PostgreSQL, code like

```rs
#[cfg(feature = "postgres")]
impl ToSql<Timestamp, diesel::pg::Pg> for Time
{
    fn to_sql(&self, out: &mut Output<diesel::pg::Pg>) -> SerResult {
	ToSql::<Timestamp, diesel::pg::Pg>::to_sql(&self.0, &mut out.reborrow())
    }
}
```

seems to work, but for SQLite it seems to be impossible to reuse the already existing the `ToSql` implementation for `NaiveDateTime`.

Trying

```rs
#[cfg(feature = "sqlite")]
impl ToSql<Timestamp, diesel::sqlite::Sqlite> for Time
{
    fn to_sql(&self, out: &mut Output<diesel::sqlite::Sqlite>) -> SerResult {
	ToSql::<Timestamp, diesel::sqlite::Sqlite>::to_sql(&self.0, mut out)
    }
}
```

fails with

```
32 |     fn to_sql(&self, out: &mut Output<diesel::sqlite::Sqlite>) -> SerResult {
   |               -----            ------------------------------
   |               |
   |               these two types are declared with different lifetimes...
33 |     ToSql::<Timestamp, diesel::sqlite::Sqlite>::to_sql(&self.0, out)
   |                                                                 ^^^ ...but data from `self` flows into `out` here
```

(`reborrow()` does not exist for sqlite backend)

**Reply by Ten0** [reply]
It looks like in this case you don't need to reborrow because you hold the value to serialize as inside your structure already so you will not encounter the lifetime issue described at https://github.com/diesel-rs/diesel/discussions/2984#discussion-3749047 .

**Reply by weiznich** [reply]
> So, to make it it complete: how do I migrate from

That's exactly one of the cases where the changelog remains true. You literally just change the signature and it work. For the sake of completeness the following impl compiles for me:

```rust


impl <DB> ToSql<Timestamp, DB> for Time
where
    DB: Backend,
    NaiveDateTime: ToSql<Timestamp, DB>
{
    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, DB>) -> diesel::deserialize::Result<IsNull> {
        self.0.to_sql(out)
    }
}

```

**Reply by ensc** [reply]
thx; the explicit `'b` lifetime makes it work again.

Fortunately, I do not have complex datatypes which require conversion in temporary values (implementing `ToSql` for sqlite seems to be impossible now without rewriting the low-level database conversion).

Due to lack of specialization in rust, it seems that the `sqlite` feature can not be active with other ones (the generic `impl <DB> ToSql<Timestamp, DB> for Time` conflicts with `impl ToSql<TimeStamp, db::sqlite::Sqlite> for Time`).  Fortunately, uninteresting for my use cases either....

**Reply by weiznich** [reply]
> Due to lack of specialization in rust, it seems that the sqlite feature can not be active with other ones (the generic impl <DB> ToSql<Timestamp, DB> for Time conflicts with impl ToSql<TimeStamp, db::sqlite::Sqlite> for Time). Fortunately, uninteresting for my use cases either....

That's just not true. The impl given above does work for all backends at the same time. It does not need any kind of specialization. For cases where you need different `ToSql` impls you can differentiate the impls based on `Backend::BindCollector`, rustc will accept that.

**Reply by ensc** [reply]
> For cases where you need different ToSql impls you can differentiate the impls based on Backend::BindCollector, rustc will accept that.

```rs
struct Foo(i16);

impl <DB> ToSql<SmallInt, DB> for Foo
where
    DB: Backend<BindCollector = diesel::query_builder::bind_collector::RawBytesBindCollector<DB>>,
    i16: ToSql<SmallInt, DB>,
{
    fn to_sql(&self, out: &mut diesel::serialize::Output<DB>) -> diesel::serialize::Result {
	let tmp = self.0 + 1;	// some conversion
	ToSql::<SmallInt, DB>::to_sql(&tmp , &mut out.reborrow())
    }
}

impl <'a, DB> ToSql<SmallInt, DB> for Foo
where
    DB: Backend<BindCollector = diesel::sqlite::connection::SqliteBindCollector<'a>>,
{
    fn to_sql(&self, out: &mut diesel::serialize::Output<DB>) -> diesel::serialize::Result {
	let tmp = self.0 + 1;	// some conversion
	out.set_value(tmp as i32);
	Ok(diesel::serialize::IsNull::No)
    }
}
```

fails here with

 - "module `connection` is private"  -> `SqliteBindCollector` does not seem to be part of the public api
 - error[E0119]: conflicting implementations of trait `diesel::serialize::ToSql<diesel::sql_types::SmallInt, _>` for type `Foo`

**Reply by Ten0** [reply]
> `SqliteBindCollector` does not seem to be part of the public api

That is on purpose: the second implementation does not need to be generic on DB, because it's only used on SQLite -> it can be written non-generically, e.g. https://docs.diesel.rs/master/diesel/serialize/trait.ToSql.html#examples, which solves both your issues.

As a side note, you have [this trick](https://stackoverflow.com/questions/40392524/conflicting-trait-implementations-even-though-associated-types-differ/40408431#40408431) that typically enables working around associated types disjointedness issues.

**Reply by ensc** [reply]
@Ten0  A specialization `impl ToSql<SmallInt, Sqlite> for Foo` was my first [try](https://gitlab-ext.sigma-chemnitz.de/ensc/sstate-server/-/blob/e3370b202b484dcd47bebf6260b5d767cd0afb52/modules/types/src/diesel/mod.rs#L24) but it failed due to the conflicting implementation error too.

I disable code now with `#[cfg(not(feature = "sqlite"))]`; it is ok for my application but for libraries which have to work with both `sqlite` and `postgres` features this will not be a solution.

**Reply by weiznich** [reply]
So I've played a bit with this and yes you cannot have one generic impl using a specific bind collector + one backend specific impl, which use a different bind collector. That seems to be a rustc limitation and something that couldn't be worked around with the associated types disjointedness trick mentioned about. The point is: This behaviour is not different than before, that's exactly what happens with diesel 1.4.x as well if you try to constrain the impls there with similar bounds. It isn't as obvious there as all default diesel backends use the same bind collector type, but it can happen there as well.

Now where do we go from here: In my opinion this issue is unfortunate but not release blocking as it can be worked around in third party crates by:

* Writing one general impl without using bind collector specific methods (By not using temporary values as part of the `to_sql` impl)
* Writing concrete impls per backend
* Using conditional compilation.

That does not mean that I will be opposed to improvements in this area, just that I by myself likely won't spend much time on improving this before cutting the next release.

**Comment by Sytten** [comment]
Thanks to all here for the information, I was totally lost when I updated my target commit.
I know v2 is not stable, but I also needed features on master and I was already targeting an older version of main.
For others, this link contains all the information: https://docs.diesel.rs/master/diesel/serialize/trait.ToSql.html#examples
